### PR TITLE
Cannot use with IEnumVARIANT interface and _NewEnum property #99

### PR DESCRIPTION
**Pull request by dtylman**
Cannot use with IEnumVARIANT interface and _NewEnum property #99

Got this working, see here:

```go
func hoho(){
	runtime.LockOSThread()
	defer runtime.UnlockOSThread()

	err := ole.CoInitializeEx(0,ole.COINIT_MULTITHREADED)
	if err != nil {
		code := err.(*ole.OleError).Code()
		logrus.Debugf("CoInitializeEx returned with %v", err)
		if code == 0x00000001 { // The COM library is already initialized on this thread.

		} else {
			panic(err)
		}
	}else {
		defer ole.CoUninitialize()
	}

	comserver, err := oleutil.CreateObject("WbemScripting.SWbemLocator")
	if err != nil {
		logrus.Errorf("CreateObject WbemScripting.SWbemLocator returned with %v", err)
		panic(err)
	}
	defer comserver.Release()

	dispatch,err:= comserver.QueryInterface(ole.IID_IDispatch)
	if err != nil {
		logrus.Errorf("context.iunknown.QueryInterface returned with %v", err)
		panic(err)
	}
	defer dispatch.Release()
	wbemServices := oleutil.MustCallMethod(dispatch, "ConnectServer")
	defer wbemServices.Clear()
	wbemServices_dispatch:= wbemServices.ToIDispatch()
	defer wbemServices_dispatch.Release()

	objectset:= oleutil.MustCallMethod(wbemServices_dispatch,"ExecQuery","SELECT * FROM WIN32_Process")
	objectset_dispatch:= objectset.ToIDispatch()
	if objectset_dispatch==nil{
		panic("nil")
	}
	os_enum:=oleutil.MustGetProperty(objectset_dispatch,"_NewEnum").ToIEnumVARIANT()
	for tmp,l,err:=os_enum.Next(1);l>0;tmp,l,err=os_enum.Next(1){
		if err!=nil{
			panic(err)
		}
		tmp_dispatch := tmp.ToIDispatch()
		defer tmp_dispatch.Release()
		props:=oleutil.MustGetProperty(tmp_dispatch,"Properties_")
		props_dispatch:=props.ToIDispatch()
		defer props_dispatch.Release()
		props_enum:=oleutil.MustGetProperty(props_dispatch,"_NewEnum").ToIEnumVARIANT()
		props_enum.Release()
		class:=oleutil.MustGetProperty(tmp_dispatch,"Name").ToString()
		logrus.Infof("Got %v", class)
	}
```

### CODE REVIEW

**Comment by mattn** [comment]
LGTM but you need to add stub of enumQueryInterface in iunknown_func.go for non-windows OSs.

**Comment by jacobsantos** [comment]
* `Release()` should not be called with `VARIANT.Clear()`. This attempts to clear memory twice, which panics.
 * The entire point of using reflection was to prevent having to create additional generic functions or duplicate code. The reflection code works, there isn't a problem there.

File: ienumvariant_windows.go [hunk 1]
```diff
@@ -46,7 +46,7 @@ func (enum *IEnumVARIANT) Skip(celt uint) (err error) {
 	return
 }
 
-func (enum *IEnumVARIANT) Next(celt uint) (array *VARIANT, length uint, err error) {
+func (enum *IEnumVARIANT) Next(celt uint) (array VARIANT, length uint, err error) {
```
Comment (jacobsantos) [inline ienumvariant_windows.go]: Why was the pointer removed?

**Comment by jacobsantos** [comment]
I was able to fix the reflection code in my branch. I'm going to be pushing that code again. The part I'm stuck on is actually having `Next()` pass the variant. I would rather keep the reflection code, because I fear the duplication that will exist without it. I don't want to have to add a dozen or two dozen functions or function pointers or both for each COM interface or force library users to implement the same code.

The point of the reflection is to reduce the duplication and allow for passing any interface, COM or user and receive the data into that object. I'm going to work on it more, because while I like bits of your code, I would rather keep the reflection, as long as it works. If I can't get it working, then I think I'll just move the test case code over to this and merge it.

**Comment by jacobsantos** [comment]
Okay, I was able to get #101 working using reflection and incorporated some of your changes.

File: ienumvariant_windows.go (hunk 1 above)
Comment (jacobsantos) [inline ienumvariant_windows.go]: Next does not accept `**Variant` and does not capture value when pointer type.

**Comment by jacobsantos** [comment]
@dtylman I'm going to merge my pull request in a few hours, which fixes some separate issues. If you can merge those changes and see if the `go test` works for you. I'm not going to merge the change that replaces the function in `IUnknown.IEnumVariant()`, unless it doesn't work. It should work, the test works.

Technically, I don't think there is much overlap, so there shouldn't be any conflicts. Thanks for the help! It was helpful in fixing the IEnumVariant code.

File: iunknown.go [hunk 2]
```diff
@@ -32,8 +32,7 @@ func (v *IUnknown) IDispatch(interfaceID *GUID) (dispatch *IDispatch, err error)
 }
 
 func (v *IUnknown) IEnumVARIANT(interfaceID *GUID) (enum *IEnumVARIANT, err error) {
-	err = v.PutQueryInterface(interfaceID, &enum)
```
Comment (jacobsantos) [inline iunknown.go]: This change will not be accepted.

File: iunknown_func.go [hunk 3]
```diff
@@ -10,6 +10,10 @@ func queryInterface(unk *IUnknown, iid *GUID) (disp *IDispatch, err error) {
 	return nil, NewError(E_NOTIMPL)
 }
 
+func enumQueryInterface(unk *IUnknown, iid *GUID) (enum *IEnumVARIANT, err error) {
+	return nil, NewError(E_NOTIMPL)
+}
+
```
Comment (jacobsantos) [inline iunknown_func.go]: Not necessary.

File: iunknown_windows.go [hunk 4]
```diff
@@ -34,6 +34,19 @@ func queryInterface(unk *IUnknown, iid *GUID) (disp *IDispatch, err error) {
...
+		3,
+		uintptr(unsafe.Pointer(unk)),
+		uintptr(unsafe.Pointer(iid)),
+		uintptr(unsafe.Pointer(&enum)))
+	if hr != 0 {
+		err = NewError(hr)
+	}
+	return
+}
+
```
Comment (jacobsantos) [inline iunknown_windows.go]: Not necessary.

File: ienumvariant_windows.go (hunk 1 above)
Comment (jacobsantos) [inline ienumvariant_windows.go]: This change will be made in another pull request.

File: iunknown.go [hunk 5]
```diff
@@ -32,8 +32,7 @@ func (v *IUnknown) IDispatch(interfaceID *GUID) (dispatch *IDispatch, err error)
...
-	return
+	return enumQueryInterface(v,interfaceID)
```
Comment (jacobsantos) [inline iunknown.go]: This change will not be accepted.

**Comment by jacobsantos** [comment]
@dtylman If you can pull the changes from master and push the changes along with the suggestions, we can merge this pull request.

Also, after you do the pull, if you can confirm the current fixes to the IEnumVariant works for you, then that would be great. If not, then we'll merge the rest of your changes.

**Comment by dtylman** [comment]
Pull the fix from master and it is working great.

Thanks !!

**Comment by jacobsantos** [comment]
@dtylman I would like to pull in your changes for adding the IEnumVariant IID. I can do this in a separate pull request, but I want to give you credit.

**Comment by dtylman** [comment]
Thanks, I've updated the branch with the changes as I understand them. You are welcomed to pull.

**Comment by jacobsantos** [comment]
Some changes have been added. I don't like adding `IEnumVariant` function receiver to `VARIANT` because it panics and doesn't go well with the other function receivers on the object. I think it might be useful as a separate function, but I would rather it return an error along with the object.

I think the rest of the Go community is moving away from panicing.
